lib/wiki_assignment_output.rb

```diff
@@ -0,0 +1,97 @@
+require './lib/wiki'
+#= Class for generating wikitext for updating assignment details on talk pages
+class WikiAssignmentOutput
+  ################
+  # Entry points #
+  ################
+  def self.build_talk_page_update(title,
+                                  talk_title,
+                                  title_assignments,
+                                  course_page)
+    initial_page_content = Wiki.get_page_content talk_title
+    # We only want to add assignment tags to non-existant talk pages if the
+    # article page actually exists. This also servces to make sure that we
+    # only post to talk pages of mainspace articles, as we assume that pages
+    # like Talk:Template:Citation or Talk:Wikipedia:Notability do not exist.
+    if initial_page_content.nil?
+      return nil if Wiki.get_page_content(title).nil?
+      initial_page_content = ''
+    end
+
+    # Limit it to live assignments for this article-course
+    title_assignments = title_assignments.select { |a| !a['deleted'] }
+
+    course_assignments_tag = assignments_tag(course_page, title_assignments)
+    page_content = build_assignment_page_content(course_assignments_tag,
+                                                 course_page,
+                                                 initial_page_content)
+    page_content
+  end
+
+  ###################
+  # Helper methods #
+  ###################
+  def self.assignments_tag(course_page, title_assignments)
+    return '' if title_assignments.empty?
+
+    # Make a list of the assignees, role 0
+    tag_assigned = build_wikitext_user_list(title_assignments, 0)
+    # Make a list of the reviwers, role 1
+    tag_reviewing = build_wikitext_user_list(title_assignments, 1)
+
+    # Build new tag
+    # NOTE: If the format of this tag gets changed, then the dashboard may
+    # post duplicate tags for the same page, unless we update the way that
+    # we check for the presense of existging tags to account for both the new
+    # and old formats.
+    dashboard_url = Figaro.env.dashboard_url
+    tag = "{{#{dashboard_url} assignment | course = #{course_page}"
+    tag += " | assignments = #{tag_assigned}" unless tag_assigned.blank?
+    tag += " | reviewers = #{tag_reviewing}" unless tag_reviewing.blank?
+    tag += ' }}'
+
+    tag
+  end
+
+  # This method creates updated wikitext for an article talk page, for when
+  # the set of assigned users for the article for a single course changes.
+  # The strategy here is to only update the tag for one course at a time, so
+  # that the user who updates the assignments for a course only introduces data
+  # for that course. We also want to make as minimal a change as possible, and
+  # to make sure that we're not disrupting the format of existing content.
+  def self.build_assignment_page_content(new_tag,
+                                         course_page,
+                                         page_content)
+
+    dashboard_url = Figaro.env.dashboard_url
```

Comment:
Using `Figaro.env` couples us to that gem. I've been meaning to bring this up. Many gems implement an interface for accessing `ENV` variables. Because `ENV['dashboard_url']` would also work, I think we should switch to that so that we could, if need be, switch to [dotenv](https://github.com/bkeepers/dotenv). Ideally we wouldn't be unnecessarily coupled to an environment gem.